src/buffer_manager.ts

```diff
@@ -622,8 +623,18 @@ export class BufferManager implements Disposable {
             this.textDocumentToBufferId.delete(document);
             unusedBuffers.push(bufId);
         });
-        unusedWindows.length && (await actions.lua("close_windows", unusedWindows));
-        unusedBuffers.length && (await actions.lua("delete_buffers", unusedBuffers));
+
+        if (unusedWindows.length || unusedBuffers.length) {
+            const toCleanup = { windows: unusedWindows, buffers: unusedBuffers };
+            // Log if cleanup takes a long time, in case the request is never fulfilled
+            const logSlowCleanup = setTimeout(() => logger.warn("Cleanup took longer than 5s: ", toCleanup), 5000);
+            // We don't await the result of this cleanup as a workaround for #2136,
+            // so that the user isn't blocked forever in case the nvim client gets stuck.
+            actions.lua("cleanup_windows_and_buffers", toCleanup).then(
+                () => clearTimeout(logSlowCleanup),
```

Comment:
nice approach. does this run on error or do we also need clearTimeout for the error case?

Reply:
Good point, moved this to a `finally` to ensure it's cleared in both cases